Mathlib/Algebra/Hom/Group.lean

```diff
@@ -0,0 +1,1698 @@
+pointwise multiplication and pointwise inversion.
+
+This file also defines the lesser-used (and notation-less) homomorphism types which are used as
+building blocks for other homomorphisms:
+
+* `ZeroHom`
+* `OneHom`
+* `AddHom`
+* `MulHom`
+* `MonoidWithZeroHom`
+
+Finally, we define classes that state the coercion operator `↑` (a.k.a. `coe`) is a homomorphism:
+ * `CoeIsOneHom`/`CoeIsZeroHom`
+ * `CoeIsMulHom`/`CoeIsAddMonoidHom`
+ * `CoeIsMonoidHom`/`CoeIsAddMonoidHom`
+ * `CoeIsMonoidWithZeroHom`
+These come with a selection of `simp` lemmas stating that `↑` preserves the corresponding operation:
+`coe_add`, `coe_mul`, `coe_zero`, `coe_one`, `coe_pow`, `coe_nsmul`, `coe_zpow`, `coe_zsmul`,
+`coe_sub`, `coe_neg`, ..., etc.
+
+## Notations
+
+* `→+`: Bundled `AddMonoid` homs. Also use for `AddGroup` homs.
+* `→*`: Bundled `Monoid` homs. Also use for `Group` homs.
+* `→*₀`: Bundled `MonoidWithZero` homs. Also use for `GroupWithZero` homs.
+* `→ₙ*`: Bundled `Semigroup` homs.
+
+## Implementation notes
+
+There's a coercion from bundled homs to fun, and the canonical
+notation is to use the bundled hom as a function via this coercion.
+
+There is no `GroupHom` -- the idea is that `MonoidHom` is used.
+The constructor for `MonoidHom` needs a proof of `map_one` as well
+as `map_mul`; a separate constructor `MonoidHom.mk'` will construct
+group homs (i.e. monoid homs between groups) given only a proof
+that multiplication is preserved,
+
+Implicit `{}` brackets are often used instead of type class `[]` brackets.  This is done when the
+instances can be inferred because they are implicit arguments to the type `MonoidHom`.  When they
+can be inferred from the type it is faster to use this method than to use type class inference.
+
+Historically this file also included definitions of unbundled homomorphism classes; they were
+deprecated and moved to `Deprecated/Group`.
+
+## Tags
+
+MonoidHom, AddMonoidHom
+
+-/
+
+
+variable {α β M N P : Type _}
+
+-- monoids
+variable {G : Type _} {H : Type _}
+
+-- groups
+variable {F : Type _}
+
+-- homs
+section Zero
+
+/-- `ZeroHom M N` is the type of functions `M → N` that preserve zero.
+
+When possible, instead of parametrizing results over `(f : ZeroHom M N)`,
+you should parametrize over `(F : Type*) [ZeroHomClass F M N] (f : F)`.
+
+When you extend this structure, make sure to also extend `ZeroHomClass`.
+-/
+structure ZeroHom (M : Type _) (N : Type _) [Zero M] [Zero N] where
+  /-- The underlying function -/
+  toFun : M → N
+  /-- The proposition that the function preserves 0 -/
+  map_zero' : toFun 0 = 0
+#align zero_hom ZeroHom
+#align zero_hom.map_zero' ZeroHom.map_zero'
+
+/-- `ZeroHomClass F M N` states that `F` is a type of zero-preserving homomorphisms.
+
+You should extend this typeclass when you extend `ZeroHom`.
+-/
+-- Porting note: `outParam` is added to instances `[Zero M] [Zero N]` until issue
+-- https://github.com/leanprover/lean4/issues/1852 is resolved
```

Comment:
This should already work, but removing the `outParam` requires a bunch of type annotations below.

Reply:
Removing the `outParam` here causes a bunch of `rw` to fail later on. For example in `of_injective`, `ZeroHomClass.map_zero f` has to be replaced by `@ZeroHomClass.map_zero F R M _ _ _ f`. Seems that the functionality is not totally working yet.